include/mrdox/Support/ExecutorGroup.hpp

```diff
@@ -0,0 +1,165 @@
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+
+#ifndef MRDOX_SUPPORT_EXECUTORGROUP_HPP
+#define MRDOX_SUPPORT_EXECUTORGROUP_HPP
+
+#include <mrdox/Platform.hpp>
+#include <mrdox/Support/ThreadPool.hpp>
+#include <mrdox/Support/unlock_guard.hpp>
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <vector>
+
+namespace clang {
+namespace mrdox {
+
+/** A set of execution agents for performing concurrent work.
+*/
+template<class Agent>
+class ExecutorGroup
+{
+    struct Impl
+    {
+        std::mutex mutex_;
```

Comment:
this seems like an application for a shared_mutex to me. Optimization ofc.

Reply:
I find the lock usage in this class confusing, but correct. Hard to read and see correctness though.